File: Platformer/scripts/Spider.jl
module SpiderModule    
    using JulGame

    mutable struct Spider <: Script
        animator
        isMovingRight::Bool
        parent::JulGame.EntityModule.Entity
        sound::JulGame.SoundSourceModule.SoundSource
        speed::EditorExport{Float64}
        startingX::EditorExport{Int}
        endingX::EditorExport{Int}

        function Spider()
            this = new()

            this.endingX = 0
            this.isMovingRight = false
            this.startingX = 0
            this.speed = 0.0

            return this
        end
    end

    function JulGame.initialize(this::Spider)
        this.animator = this.parent.animator
        this.parent.transform.position = Vector2f(this.startingX, this.parent.transform.position.y)
    end
    function JulGame.update(this::Spider, deltaTime)
        if this.parent.transform.position.x <= this.startingX && !this.isMovingRight
            JulGame.Component.flip(this.parent.sprite)
            this.isMovingRight = true
        elseif this.parent.transform.position.x >= this.endingX && this.isMovingRight
            JulGame.Component.flip(this.parent.sprite)
            this.isMovingRight = false
        end

        if this.isMovingRight
            this.parent.transform.position = JulGame.Math.Vector2f(this.parent.transform.position.x + this.speed*deltaTime, this.parent.transform.position.y)
        else
            this.parent.transform.position = JulGame.Math.Vector2f(this.parent.transform.position.x - this.speed*deltaTime, this.parent.transform.position.y)
        end
    end
end # module

File: Platformer/ts/_generated/scripts/Spider.ts
import { registerScript } from "julgame/src/engine/runtime/scriptRegistry";


    // using JulGame

    export class Spider {
        animator
        isMovingRight: boolean
        parent: Entity
        sound: SoundSource
        speed: number
        startingX: number
        endingX: number

        constructor() {
            

            this.endingX = 0
            this.isMovingRight = false
            this.startingX = 0
            this.speed = 0.0

        }
    }

    function JulGame_initialize_Spider(self: Spider) {
        self.animator = self.parent.animator
        self.parent.transform.position = {x: self.startingX, y: self.parent.transform.position.y}
    }
    function JulGame_update_Spider(self: Spider, deltaTime) {
        if (self.parent.transform.position.x <= self.startingX && !self.isMovingRight) {
            (globalThis as any).JulGame.Component.flip(self.parent.sprite)
            self.isMovingRight = true
        } else if (self.parent.transform.position.x >= self.endingX && self.isMovingRight) {
            (globalThis as any).JulGame.Component.flip(self.parent.sprite)
            self.isMovingRight = false
        }

        if (self.isMovingRight) {
            self.parent.transform.position = {x: self.parent.transform.position.x + self.speed*deltaTime, y: self.parent.transform.position.y}
        } else {
            self.parent.transform.position = {x: self.parent.transform.position.x - self.speed*deltaTime, y: self.parent.transform.position.y}
        }
    }
registerScript("Spider", {
    create: () => new Spider(),
    initialize: JulGame_initialize_Spider,
    update: JulGame_update_Spider,
});
